django/forms/fields.py

```diff
@@ -764,7 +763,14 @@ def __init__(self, choices_func):
         self.choices_func = choices_func
 
     def __iter__(self):
-        yield from self.choices_func()
+        value = self.choices_func()
+        if isinstance(value, dict):
+            value = (value,)
+        for item in value:
+            if isinstance(item, dict):
+                yield from ChoiceField._flatten_dictionary_choices(item)
+            else:
+                yield item
```

Comment:
This bit seems like a shame... — what's the case that calls for the complication? (A dict isn't callable... 🤔)

Reply:
The issue stems from dictionaries being iterable. If `choices_func()` returns a dict, then `yield from` will only yield the keys.

On top of that there is nothing to stop `choices_func()` returning an _iterable of dicts_ (if it's a generator for example), so we check each value we iterate over.

I've added a comment explaining this, and replaced `dict` with `Mapping`.